File: frameKT/src/main/java/com/guoliang/framekt/util/DensityUtil.java
package com.guoliang.framekt.util;

import android.content.res.Resources;
import android.util.DisplayMetrics;
import android.util.TypedValue;

/**
 * @Description:
 * @Author: zhangguoliang
 * @CreateTime: 2020/5/26 18:08
 */
public class DensityUtil {

    private static Resources sResources=Resources.getSystem();

    public static float dip2px(float dp){
        DisplayMetrics dm = sResources.getDisplayMetrics();
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP,dp,dm);
    }

    public static int dip2pxInt(float dp){
        return (int) dip2px(dp);
    }

    public static float sp2px(float sp){
        DisplayMetrics dm = sResources.getDisplayMetrics();
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP,sp,dm);
    }

    public static int sp2pxInt(float sp){
        return (int) sp2px(sp);
    }

    /**
     * 根据手机的分辨率PX(像素)转成DP
     * @param pxValue
     * @return
     */
    public static int px2dip(Float pxValue) {
        float scale = sResources.getDisplayMetrics().density;
        return (int) (pxValue/scale+0.5f);
    }

    /**
     * 将px值转换为sp值，保证文字大小不变
     * @param pxValue
     * @return
     */
    public static int px2sp(Float pxValue) {
        float fontScale = sResources.getDisplayMetrics().scaledDensity;
        return (int) (pxValue/fontScale+0.5f);
    }
}

File: frameKT/src/main/java/com/guoliang/framekt/util/DensityUtil.kt
package com.guoliang.framekt.util

import android.content.res.Resources
import android.util.TypedValue

/**
 * @Description:
 * @Author: zhangguoliang
 * @CreateTime: 2020/5/26 18:08
 */
object DensityUtil {
    private val sResources = Resources.getSystem()
    fun dip2px(dp: Float): Float {
        val dm = sResources.displayMetrics
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, dm)
    }

    fun dip2pxInt(dp: Float): Int {
        return dip2px(dp).toInt()
    }

    fun sp2px(sp: Float): Float {
        val dm = sResources.displayMetrics
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, sp, dm)
    }

    fun sp2pxInt(sp: Float): Int {
        return sp2px(sp).toInt()
    }

    /**
     * 根据手机的分辨率PX(像素)转成DP
     * @param pxValue
     * @return
     */
    fun px2dip(pxValue: Float): Int {
        val scale = sResources.displayMetrics.density
        return (pxValue / scale + 0.5f).toInt()
    }

    /**
     * 将px值转换为sp值，保证文字大小不变
     * @param pxValue
     * @return
     */
    fun px2sp(pxValue: Float): Int {
        val fontScale = sResources.displayMetrics.scaledDensity
        return (pxValue / fontScale + 0.5f).toInt()
    }
}